### PR TITLE
Randomize odometry

### PR DESCRIPTION
**Pull request by ThijsBeurskensTUE**
set a random odometry starting position when uncertain_odom is set to true

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -115,15 +115,15 @@ int main(int argc, char **argv){
     std::vector<RobotPtr> robots;
     geo::Pose3D spawnLocation = config.spawn.value();
     Id pyro_id = world.addObject(spawnLocation, robot_shape, robot_color, robottype);
-    robots.push_back(std::make_shared<Robot>("pyro", pyro_id));
-    robots.back()->base.setDisableSpeedCap(config.disable_speedcap.value());
-    robots.back()->base.setUncertainOdom(config.uncertain_odom.value());
+    robots.push_back(std::make_shared<Robot>("pyro", pyro_id, config.disable_speedcap.value(), config.uncertain_odom.value()));
+    //robots.back()->base.setDisableSpeedCap(config.disable_speedcap.value());
```
Comment (KdVos) [inline src/main.cpp:119]:
```suggestion

```

File: src/main.cpp [hunk 2]
```diff
@@ -115,15 +115,15 @@ int main(int argc, char **argv){
...
+    //robots.back()->base.setUncertainOdom(config.uncertain_odom.value());
```
Comment (KdVos) [inline src/main.cpp:120]:
```suggestion
```

File: src/main.cpp [hunk 3]
```diff
@@ -115,15 +115,15 @@ int main(int argc, char **argv){
...
 
     if (config.enable_taco.value()){
         Id taco_id = world.addObject(spawnLocation, robot_shape, robot_color, robottype);
-        robots.push_back(std::make_shared<Robot>("taco", taco_id));
-        robots.back()->base.setDisableSpeedCap(config.disable_speedcap.value());
-        robots.back()->base.setUncertainOdom(config.uncertain_odom.value());
+        robots.push_back(std::make_shared<Robot>("taco", taco_id, config.disable_speedcap.value(), config.uncertain_odom.value()));
+        //robots.back()->base.setDisableSpeedCap(config.disable_speedcap.value());
```
Comment (KdVos) [inline src/main.cpp:125]:
```suggestion
```

File: src/main.cpp [hunk 4]
```diff
@@ -115,15 +115,15 @@ int main(int argc, char **argv){
...
+        //robots.back()->base.setUncertainOdom(config.uncertain_odom.value());
```
Comment (KdVos) [inline src/main.cpp:126]:
```suggestion
```

File: src/robot.cpp [hunk 5]
```diff
@@ -20,7 +20,7 @@ void Robot::speakCallback(const std_msgs::String::ConstPtr& msg)
 
 // ----------------------------------------------------------------------------------------------------
 
-Robot::Robot(std::string name, Id id)
+Robot::Robot(std::string name, Id id, bool disable_speedcap, bool uncertain_odom) : base(disable_speedcap, uncertain_odom)
```
Comment (KdVos) [inline src/robot.cpp:23]:
```suggestion
Robot::Robot(std::string name, Id id, bool disable_speedcap = false, bool uncertain_odom= false) : base(disable_speedcap, uncertain_odom)
```

File: src/virtualbase.h [hunk 6]
```diff
@@ -24,27 +25,38 @@ class Virtualbase
...
-
         std::random_device rd;
         gen = std::mt19937(rd());
         //std::normal_distribution<double> dis(0.0,0.003);
         dis = std::uniform_real_distribution<double>(-0.002,0.002);
 
-        updateWheelUncertaintyFactors();
-    }
+        if (uncertain_odom_) {
+            std::uniform_real_distribution<double> dis2 = std::uniform_real_distribution<double>(-2.0,2.0);
```
Comment (KdVos) [inline src/virtualbase.h:36]:
```suggestion
            std::uniform_real_distribution<double> dis_pos_init = std::uniform_real_distribution<double>(-2.0,2.0);
```

File: src/virtualbase.h [hunk 7]
```diff
@@ -24,27 +25,38 @@ class Virtualbase
...
+            odometry_state.pose.pose.position.x = dis2(gen);
+            odometry_state.pose.pose.position.y = dis2(gen);
+            odometry_state.pose.pose.position.z = 0.0;
+            dis2 = std::uniform_real_distribution<double>(0.0, 2 * M_PI);
```
Comment (KdVos) [inline src/virtualbase.h:40]:
```suggestion
             std::uniform_real_distribution<double> dis_angle_init = std::uniform_real_distribution<double>(0.0, 2 * M_PI);
```

**Review by KdVos — CHANGES_REQUESTED** [review]

File: src/robot.cpp (hunk 5 above)
Comment (MatthijsBurgh) [reply src/robot.cpp:23]: No, default values should be ONLY defined in the header.

File: src/virtualbase.h [hunk 8]
```diff
@@ -24,27 +25,38 @@ class Virtualbase
 {
 
 public:
-    Virtualbase()
+    Virtualbase(bool disable_speedcap = false, bool uncertain_odom = false) : disable_speedcap_(disable_speedcap), uncertain_odom_(uncertain_odom)
```
Comment (MatthijsBurgh) [inline src/virtualbase.h:28]: Move definitions to source  file...

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

File: src/virtualbase.h [hunk 9]
```diff
@@ -8,6 +8,7 @@
 #include "geometry_msgs/Twist.h"
 #include "geometry_msgs/Pose.h"
 #include "nav_msgs/Odometry.h"
+#include "tf/transform_datatypes.h"
```
Comment (MatthijsBurgh) [inline src/virtualbase.h:11]: No, get rid of all `tf` stuff. Replace it by all `tf2`.

File: src/robot.h [hunk 10]
```diff
@@ -6,6 +6,7 @@
 #include "world.h"
 
 // ROS
+#include <ros/ros.h>
 #include <ros/publisher.h>
```
Comment (MatthijsBurgh) [inline src/robot.h:10]:
```suggestion
#include <ros/node_handle.h>
#include <ros/publisher.h>
#include <ros/subscriber.h>
```

File: src/robot.cpp (hunk 5 above)
Comment (MatthijsBurgh) [inline src/robot.cpp:23]: Use const ref for `name` arg

File: src/virtualbase.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
+#include "virtualbase.h"
```
Comment (MatthijsBurgh) [inline src/virtualbase.cpp:1]: Move includes from header to source file for includes that are only used in the source file

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]
See comments

**Review by KdVos — APPROVED** [review]

File: src/virtualbase.h [hunk 12]
```diff
@@ -8,9 +8,9 @@
 #include "geometry_msgs/Twist.h"
 #include "geometry_msgs/Pose.h"
 #include "nav_msgs/Odometry.h"
-#include "tf/tf.h"
 
-#include "random"
+#include <random>
+#include <cmath>
```
Comment (MatthijsBurgh) [inline src/virtualbase.h:13]: This one can be moved to the source file.

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]

**Review by KdVos — APPROVED** [review]

**Pull request merged**